pkg/build/controller/buildimagechangetrigger/buildimagechangetrigger_controller.go

```diff
@@ -0,0 +1,287 @@
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildImageChangeTriggerController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	is, err := bc.getImageStreamByKey(key.(string))
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+
+	if is == nil {
+		return false
+	}
+
+	err = bc.HandleImageStream(is)
+	bc.handleError(err, key, is)
+
+	return false
+
+}
+
+func (bc *BuildImageChangeTriggerController) getImageStreamByKey(key string) (*imageapi.ImageStream, error) {
+	obj, exists, err := bc.isStore.GetByKey(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to retrieve imagestream %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return nil, err
+	}
+	if !exists {
+		glog.V(4).Infof("ImageStream %q is no longer in the shared informer store to imagestream lister cache (perhaps it has been deleted)", key)
+		return nil, nil
+	}
+
+	return obj.(*imageapi.ImageStream), nil
+}
+
+func (bc *BuildImageChangeTriggerController) HandleImageStream(stream *imageapi.ImageStream) error {
+	glog.V(4).Infof("Build image change controller detected ImageStream change for %s/%s and docker image respository %s", stream.Namespace, stream.Name, stream.Status.DockerImageRepository)
+	hasError := false
+
+	bcs, err := bc.bcStore.GetConfigsForImageStreamTrigger(stream.Namespace, stream.Name)
+	if err != nil {
+		return err
+	}
+
+	for _, config := range bcs {
+		var (
+			from            *kapi.ObjectReference
+			shouldBuild     = false
+			triggeredImage  = ""
+			latestReference string
+		)
+		// For every ImageChange trigger find the latest tagged image from the image stream and
+		// invoke a build using that image id. A new build is triggered only if the latest tagged image id or pull spec
+		// differs from the last triggered build recorded on the build config for that trigger
+		for _, trigger := range config.Spec.Triggers {
+			if trigger.Type != buildapi.ImageChangeBuildTriggerType {
+				continue
+			}
+			if trigger.ImageChange.From != nil {
+				from = trigger.ImageChange.From
+			} else {
+				from = buildutil.GetInputReference(config.Spec.Strategy)
+			}
+
+			if from == nil || from.Kind != "ImageStreamTag" {
+				continue
+			}
+			fromStreamName, tag, ok := imageapi.SplitImageStreamTag(from.Name)
+			if !ok {
+				glog.Errorf("Invalid image stream tag: %s in build config %s/%s", from.Name, config.Name, config.Namespace)
+				continue
+			}
+
+			fromNamespace := from.Namespace
+			if len(fromNamespace) == 0 {
+				fromNamespace = config.Namespace
+			}
+
+			// only trigger a build if this image stream matches the name and namespace of the stream ref in the build trigger
+			// also do not trigger if the imagerepo does not have a valid DockerImageRepository value for us to pull
+			// the image from
+			if len(stream.Status.DockerImageRepository) == 0 || fromStreamName != stream.Name || fromNamespace != stream.Namespace {
```

Comment:
I would think that the only case that stream.Status.DockerImageRepository is not populated is that there's no registry configured in the cluster. @bparees why do we check that here?